### PR TITLE
Change Set to implement Hash directly to reduce memory usage.

### PR DESCRIPTION
**Pull request by mrjbq7**
I thought it might be a good idea to have a ``Set`` implementation that uses less memory (by not storing a ``nil`` value in each ``Entry``).

A simple performance test shows an advantage (although primarily the advantage should be in less memory used):

```crystal
s = Set(Int32).new
1000.times do
  (0...100000).each do |v|
    s.add v
  end
end
puts s.length
```

Crystal 0.7.4 [42efdbf] (Wed Jun 24 00:55:53 UTC 2015)

```
$ crystal build --release perf_set.cr
$ time ./perf_set 
100000

real	0m0.820s
user	0m0.813s
sys	0m0.008s
```

Crystal master (e19dd68d3151128834e27992f2b84774e242738a)

```
$ bin/crystal build --release perf_set.cr
Using compiled compiler at .build/crystal
$ time ./perf_set 
100000

real	0m0.770s
user	0m0.759s
sys	0m0.008s
```

The parts of ``hash.cr`` that were used were copied directly and modified as necessary to remove the ``value``.

### CODE REVIEW

**Comment by asterite** [comment]
@mrjbq7 This is an **excellent** observation!

Now, wouldn't it be *really* cool if you didn't have to do these kind of optimizations by hand, and we had a smarter compiler?

In fact, several days ago @waj made me the same comment: do nil instance variables occupy space in the final object? And I told him that yes, because we represent nil as i1, the smallest integer (1 bit). But it was a comment made far from our computers so we probably forgot about it. So now that you bring this subject again, I searched for "llvm empty type" and it turns out there is such thing: an empty struct!

So, there are actually two things here:

1. Make nil instance variables not occupy space in an object.
2. Make Set a struct. Because Set is immutable (it creates a hash but later doesn't modify that hash reference, just the hash itself) it could perfectly be a struct. This is what structs are for: create lightweight wrappers around other objects, so you don't allocate extra memory.

However, for point 2 some days ago I changed Set to a struct and run the [havlak](https://github.com/manastech/crystal/blob/master/samples/havlak.cr) benchmark and it turned out to be slower that way, I don't know why. But today I try it and in fact it turns out to be faster, so I must have been doing something wrong...

Well, as you can see, I just made two commits for each of the points above. And because point 1 is a general optimization to the code generated by the compiler, everyone wins! Even the compiler!

The time it takes to compile the compiler went from 16.6s / 8.74s down to 14.85s / 6.78s . There are two times because one is the first run where nothing is cached, the second is the one using cached information from the previous compilation.

Time for the havlak benchmark went from 12.45s to 11.13s, and I'm sure the memory usage went down to but I didn't measure it.

In short: **BIG THANKS!!** These are the kind of remarks/comments that the language needs :-)

And, of course, the time for your snippet is now faster, from 3.03s to 2.5s (I tried with 4000 iterations, it's better to see these differences with bigger numbers).

Small note: you don't need to compile the compiler to try changes in the standard library. If you use the `bin/crystal` wrapper, that will use the "new" (current directory) std.

**Comment by asterite** [comment]
@mrjbq7 I guess we can close this pull request?

**Comment by mrjbq7** [comment]
@asterite Wow, awesome quick fix!  It's always frustrating to duplicate code - I didn't know the optimizations worked the way you describe - so very cool.  Thanks for the explanation!
